pyomo/contrib/pyros/master_problem_methods.py

```diff
@@ -0,0 +1,344 @@
+
+    if config.solve_master_globally:
+        backup_solvers = deepcopy(config.backup_global_solvers)
+    else:
+        backup_solvers = deepcopy(config.backup_local_solvers)
+    backup_solvers.insert(0, solver)
+
+    if not solver.available():
+        raise RuntimeError("NLP solver %s is not available." %
+                           config.solver)
+
+    higher_order_decision_rule_efficiency(config, model_data)
+
+    while len(backup_solvers) > 0:
+        solver = backup_solvers.pop(0)
+        try:
+            results = solver.solve(nlp_model, tee=config.print_subsolver_progress_to_screen)
+        except ValueError as err:
+            if 'Cannot load a SolverResults object with bad status: error' in str(err):
+                results.solver.termination_condition = tc.error
+                results.solver.message = str(err)
+                master_soln.results = results
+                master_soln.grcs_termination_condition = grcsTerminationCondition.subsolver_error
+                return master_soln, ()
+            else:
+                raise
+        solver_term_cond_dict[str(solver)] = str(results.solver.termination_condition)
+        master_soln.termination_condition = results.solver.termination_condition
+        master_soln.grcs_termination_condition = None  # determined later in the algorithm
+        master_soln.fsv_vals = list(v.value for v in nlp_model.scenarios[0, 0].util.first_stage_variables)
+
+        if config.objective_focus is ObjectiveType.nominal:
+            master_soln.ssv_vals = list(v.value for v in nlp_model.scenarios[0, 0].util.second_stage_variables)
+            master_soln.second_stage_objective = value(nlp_model.scenarios[0, 0].second_stage_objective)
+        else:
+            idx =  max(nlp_model.scenarios.keys())[0]
+            master_soln.ssv_vals = list(v.value for v in nlp_model.scenarios[idx, 0].util.second_stage_variables)
+            master_soln.second_stage_objective = value(nlp_model.scenarios[idx, 0].second_stage_objective)
+        master_soln.first_stage_objective = value(nlp_model.scenarios[0, 0].first_stage_objective)
+
+        master_soln.nominal_block = nlp_model.scenarios[0, 0]
+        master_soln.results = results
+        master_soln.master_model = nlp_model
+
+        master_soln.master_subsolver_results = process_termination_condition_master_problem(config=config, results=results)
+
+        if master_soln.master_subsolver_results[0] == False:
+            return master_soln
+
+    # === At this point, all sub-solvers have been tried and none returned an acceptable status or return code
+    path = os.path.dirname(os.path.realpath(__file__))
+    save_dir = os.path.join(path, "failed_models")
+    if not os.path.exists(save_dir):
+        os.makedirs(save_dir)
+    name = os.path.join(save_dir,  config.uncertainty_set.type + "_" + model_data.original.name + "_master_" + str(model_data.iteration) + ".gms")
+    nlp_model.write(name, format="gams")
```

Comment:
We would like to have some functionality wherein subproblems that don't solve as expected get written to a gams file (or some other model file) so that the user has the ability to debug or forward to the development team (me). Is there a way we can do this, or is it not preferred Pyomo behavior?

Reply:
I would add a user option (something like “output_failed_subproblems_to” or “subproblem_export_path”?) that defaults to None. If the user specifies a path, we will dump the models to that location (but only if the user sets the path). That gives maximum flexibility to the user and makes no assumptions about the writeability of any directory.